src/v/archival/archiver_operations_impl.cc

```diff
@@ -61,12 +88,108 @@ class archiver_operations_impl : public archiver_operations_api {
       , _bucket(std::move(bucket))
       , _sg(sg) {}
 
-    /// Return upload candidate(s) if data is available or not_enough_data error
-    /// if there is not enough data to start an upload.
+    /// Return upload candidate(s) if data is available or not_enough_data
+    /// error if there is not enough data to start an upload.
     ss::future<result<reconciled_upload_candidates_list>>
     find_upload_candidates(
       retry_chain_node& workflow_rtc,
       upload_candidate_search_parameters arg) noexcept override {
+        vlog(_rtclog.debug, "find_upload_candidates {}", arg);
+        try {
+            auto partition = _pm->get_partition(arg.ntp);
+            if (partition == nullptr) {
+                // maybe race condition (partition was stopped or moved)
+                vlog(
+                  _rtclog.debug,
+                  "find_upload_candidates - can't find partition {}",
+                  arg.ntp);
+                co_return error_outcome::unexpected_failure;
+            }
+            reconciled_upload_candidates_list result{
+              .ntp = arg.ntp,
+              .read_write_fence = partition->get_applied_offset(),
+            };
+            auto base_offset = model::next_offset(
+              partition->get_uploaded_offset());
+
+            // If not limit is specified (quota is set to nullopt) we don't want
+            // to upload unlimited amount of data in one iteration. The defaults
+            // in this case are configured to allow only three segments to be
+            // uploaded.
+            auto size_quota = arg.upload_size_quota.value_or(
+              arg.target_size * 3);
+            // up to 3 requests to upload one segment
+            constexpr size_t default_req_quota = 12;
```

Comment:
nit: it's still unclear from this comment where 12 is coming from.

I think there's a missing comment somewhere explaining that we are going to upload up to 4 segments at a time

Reply:
fixed